# Chapter 4 - Exercise 1: Thực hiện những yêu cầu liên quan đến series, và đối chiếu với kết quả Output

## Part 1: Thực hiện các phép toán trên series

In [404]:
import numpy as np
import pandas as pd

In [405]:
# Câu 1a: Tạo arr_1 có 5 phần tử số nguyên chẵn 2-4-6-8-10 , arr_2 có 5 phần tử số nguyên lẻ 1-3-5-7-11
arr_1 = np.array([2, 4, 6, 8, 10])
arr_2 = np.array([1, 3, 5, 7, 11])

# In danh sách các phần tử của arr_1 và arr_2
ser1 = pd.Series(arr_1)
ser2 = pd.Series(arr_2)
print (ser1)
print (ser2)

0     2
1     4
2     6
3     8
4    10
dtype: int64
0     1
1     3
2     5
3     7
4    11
dtype: int64


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  
  <pre>0     2
1     4
2     6
3     8
4    10
dtype: int32

0     1
1     3
2     5
3     7
4    11
dtype: int32
</pre>
</details>

In [406]:
# Câu 1b: Thực hiện phép toán và thể hiện kết quả của:  ser1 + ser2 

print(ser1 + ser2)

0     3
1     7
2    11
3    15
4    21
dtype: int64


<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>0     3
1     7
2    11
3    15
4    21
dtype: int32</pre>

</details>

In [407]:
# Câu 1c: Thực hiện phép toán và thể hiện kết quả của:  ser1 - ser2
print (ser1 - ser2)

0    1
1    1
2    1
3    1
4   -1
dtype: int64


In [408]:
# Câu 1d: Thực hiện phép toán và thể hiện kết quả của:  ser1 * ser2
print (ser1 * ser2)

0      2
1     12
2     30
3     56
4    110
dtype: int64


In [409]:
# Câu 1e: Thực hiện phép toán và thể hiện kết quả của:  ser1 / ser2
print (ser1 / ser2)

0    2.000000
1    1.333333
2    1.200000
3    1.142857
4    0.909091
dtype: float64


<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>0    2.000000
1    1.333333
2    1.200000
3    1.142857
4    0.909091
dtype: float64</pre>

</details>

In [410]:
# Câu 2a: Kiểm tra xem các phần tử của ser1 có > các phần tử của ser2 không?
print (ser1.gt(ser2))

0     True
1     True
2     True
3     True
4    False
dtype: bool


<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>0     True
1     True
2     True
3     True
4    False
dtype: bool</pre>

</details>

In [411]:
# Câu 2b: Kiểm tra xem các phần tử của ser1 có < các phần tử của ser2 không?
print (ser1.lt(ser2))

0    False
1    False
2    False
3    False
4     True
dtype: bool


In [412]:
# Câu 2c: Kiểm tra xem các phần tử của ser1 có = các phần tử của ser2 không?
print (ser1.eq(ser2))

0    False
1    False
2    False
3    False
4    False
dtype: bool


In [413]:
# Câu 3a: Thêm 2 phần tử [6, 12] vào ser2
ser2 = ser2.append(pd.Series([6, 12]))
print(ser2)

0     1
1     3
2     5
3     7
4    11
0     6
1    12
dtype: int64


<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>0     1
1     3
2     5
3     7
4    11
0     6
1     12
dtype: int64
</pre>
</details>

In [414]:
# Câu 3b: Tạo series ser3 chỉ chứa các phần tử có trong ser1 mà không có trong ser2. 

ser2.reset_index(inplace=True, drop=True)
ser3 = ser1[~ser1.isin(ser2)]

#         In danh sách các phần tử của 
print(ser3)

0     2
1     4
3     8
4    10
dtype: int64


<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>0     2
1     4
3     8
4    10
dtype: int32
</pre>

</details>

In [415]:
# Câu 3c: Tạo series ser4 chỉ chứa các phần tử có trong ser2 mà không có trong ser1. 
ser4 = ser2[~ser2.isin(ser1)]

#         In danh sách các phần tử của ser4
print(ser4)

0     1
1     3
2     5
3     7
4    11
6    12
dtype: int64


<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>0     1
1     3
2     5
3     7
4    11
1    12
dtype: int64
</pre>

</details>

In [416]:
# Câu 4: Tạo series ser5 chứa các phần tử chỉ có trong ser1 và chỉ có trong ser2
ser5 = pd.concat([ser3, ser4])
#        In danh sách các phần tử của ser5
print(ser5)

0     2
1     4
3     8
4    10
0     1
1     3
2     5
3     7
4    11
6    12
dtype: int64


<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>0     2
1     4
3     8
4    10
0     1
1     3
2     5
3     7
4    11
1    12
dtype: int64
</pre>

</details>

## Part 2: Truy xuất các phần tử, và thống kê thông tin trên series

In [417]:
np.random.seed(1)
# Câu 1a: Tạo series ser6 có 35 phần tử số nguyên ngẫu nhiên có giá trị trong khoảng từ 1 đến 9. 
ser6 = pd.Series(np.random.randint(1, 9, size=35))
print(ser6)

0     6
1     4
2     5
3     1
4     8
5     2
6     4
7     6
8     8
9     1
10    1
11    2
12    5
13    8
14    6
15    5
16    7
17    2
18    3
19    5
20    7
21    6
22    3
23    5
24    4
25    5
26    3
27    5
28    6
29    7
30    3
31    5
32    8
33    8
34    2
dtype: int64


<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>(35,)

0    6
1    9
2    6
3    1
4    1
dtype: int32

30    8
31    4
32    7
33    6
34    2
dtype: int32
</pre>

</details>

In [418]:
# Câu 1b: In danh sách các phần tử của ser6 theo dạng array
print(ser6.values)

[6 4 5 1 8 2 4 6 8 1 1 2 5 8 6 5 7 2 3 5 7 6 3 5 4 5 3 5 6 7 3 5 8 8 2]


<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>[6 9 6 1 1 2 8 7 3 5 6 3 5 3 5 8 8 2 8 1 7 8 7 2 1 2 9 9 4 9 8 4 7 6 2]
</pre>

</details>

In [419]:
# Câu 1c: Cho biết thông tin thống kê chung (describe()) của ser6
ser6.describe()

count    35.000000
mean      4.742857
std       2.187435
min       1.000000
25%       3.000000
50%       5.000000
75%       6.000000
max       8.000000
dtype: float64

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>count    35.000000
mean      5.200000
std       2.763204
min       1.000000
25%       2.500000
50%       6.000000
75%       8.000000
max       9.000000
dtype: float64</pre>

</details>

In [420]:
# Câu 1d: Cho biết tổng của các phần tử có trong ser6
print(ser6.sum())

166


<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre> 182 </pre>
</details>

In [421]:
# Câu 1e: Cho biết phần tử có tần suất xuất hiện nhiều nhất trong ser6

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>0    8
dtype: int32</pre>

</details>

In [422]:
# Câu 2:  Liệt kê các dòng trong ser6 mà giá trị chia hết cho 2 và cho 3
print(ser6[(ser6 % 2 == 0) & (ser6 % 3 == 0)])

0     6
7     6
14    6
21    6
28    6
dtype: int64


<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>0     6
2     6
10    6
33    6
dtype: int32</pre>

</details>

In [423]:
# Câu 3: In các phần tử ở vị trí 0, 5, 10, 15 trong ser6
print(ser6[:20:5])
print(ser6[[0,6,10,15]])

0     6
5     2
10    1
15    5
dtype: int64
0     6
6     4
10    1
15    5
dtype: int64


<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>0     6
5     2
10    6
15    8
dtype: int32</pre>

</details>

In [424]:
# Câu 4: In ra các giá trị unique (array) trong ser6
ser6.unique()

array([6, 4, 5, 1, 8, 2, 7, 3])

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>array([6, 9, 1, 2, 8, 7, 3, 5, 4], dtype=int64)</pre>
</details>

In [425]:
# Câu 5: Tạo series ser7 với mỗi phần tử có giá trị = lập phương của phần tử trong ser6. 
ser7 = ser6.map(lambda x: x^3)
print(ser7)

0      5
1      7
2      6
3      2
4     11
5      1
6      7
7      5
8     11
9      2
10     2
11     1
12     6
13    11
14     5
15     6
16     4
17     1
18     0
19     6
20     4
21     5
22     0
23     6
24     7
25     6
26     0
27     6
28     5
29     4
30     0
31     6
32    11
33    11
34     1
dtype: int64


<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>0    216.0
1    729.0
2    216.0
3      1.0
4      1.0
dtype: float64
</pre>

</details>

## Part 3: Tạo series từ list, chuỗi và biểu thức điều kiện 

Hướng dẫn chung:
* Lập biểu thức điều kiện: <span style="color:blue">biến_series.map(lambda)</span>

In [426]:
# Câu 1: Cho list sau:
lst = ["abc", "defg", "htlmj", "dfg", "ljsac"]

In [427]:
# Câu 1: Tạo series từ list này và tạo series ser_dodai với mỗi phần tử có giá trị là chiều dài của mỗi phần tử trong ser_chuoi
ser_chuoi = pd.Series(lst)
ser_dodai = pd.Series(ser_chuoi.map(lambda x: len(x)))
print(ser_dodai)

0    3
1    4
2    5
3    3
4    5
dtype: int64


<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>0    3
1    4
2    5
3    3
4    5
dtype: int64</pre>

</details>

In [428]:
import math
def isPrime (num):
    if (num < 2):
        return False
    else:
        counter = 2
        counter_guard = 0
        while counter <= math.sqrt(num):
            if num % counter == 0:
                counter_guard += 1
            counter += 1
        return True if counter_guard == 0 else False

In [429]:
# Câu 2: Cho ser = pd.Series(np.array([1, 2, 4, 5, 8, 7, 6, 9])). 
# Sử dụng biểu thức điều kiện thích hợp để in ra các dòng trong ser có giá trị là số nguyên tố
ser = pd.Series(np.array([1, 2, 4, 5, 8, 7, 6, 9]))
print(ser[ser.map(lambda x: isPrime(x))])

1    2
3    5
5    7
dtype: int64


<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>1    2
3    5
5    7
dtype: int32
</pre>

</details>

In [430]:
# Câu 3: Cho mẫu email như sau:
pattern ='[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,4}'

# Tạo một series ser_ch, với mỗi phần tử trong ser_ch là một chuỗi
# Gợi ý: 'reading newspaper from tuoitre.vn', 'tubirona@gmail.com', 'nguyen.nn@yahoo.com', 'tran_2014@hotmail.com.vn'
ser_ch = pd.Series(np.array(['reading newspaper from tuoitre.vn', 'tubirona@gmail.com', 'nguyen.nn@yahoo.com', 'tran_2014@hotmail.com.vn']))

# In ra những dòng trong ser_ch thỏa điều kiện chuỗi là email
print(ser_ch[ser_ch.str.match(pattern)])

1          tubirona@gmail.com
2         nguyen.nn@yahoo.com
3    tran_2014@hotmail.com.vn
dtype: object


<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>1          tubirona@gmail.com
2         nguyen.nn@yahoo.com
3    tran_2014@hotmail.com.vn
dtype: object</pre>

</details>

In [431]:
# Câu 4: Cho series:
ser_names = pd.Series(['Manufacturer', 'Model', 'CarType', 'Min_Price', 'Price', 'Max_Price',
       'MPG_city', 'MPG_highway', 'AirBags', 'DriveTrain', 'Cylinders',
       'EngineSize', 'Horsepower', 'RPM', 'Rev_per_mile', 'Man_trans_avail',
       'Fuel_tank_capacity', 'Passengers', 'Length', 'Wheelbase', 'Width',
       'Turn_circle', 'Rear_seat_room', 'Luggage_room', 'Weight', 'Origin',
       'Make'])
# Sử dụng biểu thức điều kiện thích hợp để in ra các dòng của ser_names thỏa điều kiện trong chuỗi có chữ 'Price'
print(ser_names[ser_names.str.contains('Price')])

3    Min_Price
4        Price
5    Max_Price
dtype: object


<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
<pre>3    Min_Price
4        Price
5    Max_Price
dtype: object
</pre>

</details>